### PR TITLE
PrivacyPolicyBanner: remove the component and associated feature flag

### PR DESCRIPTION
**Pull request by jsnajdr**
**Update:** We decided to remove the `PrivacyPolicyBanner` completely, as it hasn't been used in the last 4 years and is unlikely to be used again.

This PR rewrites the `PrivacyPolicyBanner` component so that it uses React Query to do its `/privacy-policy` REST request, instead of using the legacy `http-data` framework. It also makes it a functional component, meaning that the PR is complete rewrite 🙂 

After the rewrite, I finally hide it completely behind a disabled `privacy-policy` flag. Last time the component was active was 4 years ago, in January 2018, when we had a privacy policy update that required getting a new consent from all users.

Since then, every load of Calypso Stats page downloads 100kB of data from the `/privacy-policy` endpoint, where the response contains two complete copies of the policy text, in HTML and plain text versions, only to discover it doesn't need to do anything about it.

If anyone ever changes the privacy policy version or the notification period, they'll need to toggle the flag back to `true`.

**How to test:**
Go to your site's stats page with some feature flags manually enabled:
```
http://calypso.localhost:3000/stats/day/example.blog?flags=privacy-policy,privacy-policy/test
```
and verify that you see a privacy policy banner:

Screenshot 2022-01-06 at 14 56 29

and that clicking on "Learn More" opens a modal dialog with the full policy text.

### CODE REVIEW

**Review by noahtallen — COMMENTED** [review]
Code changes look great to me! I did notice one possible bug. If I wait long enough with the modal open, it appears to close and then reopen. See the 16/17 second mark below: (Note that I did not click outside of the modal or do anything to trigger it)

https://user-images.githubusercontent.com/6265975/148602701-7d271ca9-2bb0-4ad2-8838-9bc9a7dd3132.mp4

I'm also curious: do we know that all users have accepted the privacy policy so that we can turn it off? At what point would someone decide to turn it off again after enabling it in the future?

**Comment by jsnajdr** [comment]
> do we know that all users have accepted the privacy policy so that we can turn it off? At what point would someone decide to turn it off again after enabling it in the future?

The banner was displayed only during the so called `notification_period`, specified by the REST response:

Screenshot 2022-01-10 at 13 35 16

If a user didn't accept it during the notification period, we didn't show the banner any more.

I'm not sure how would it work today, most likely the rules are more strict. Note that the period ended in January 2018, several months before GDPR started being enforceable in May 2018.

**Comment by jsnajdr** [comment]
> If I wait long enough with the modal open, it appears to close and then reopen.

This is a weird behavior, I can't reproduce it but have one theory: it's caused by the `isFetchingPreferences` selector returning `true` for a few seconds, while some other component fetches preferences in background. Then `shouldRender()` returns `false`, hiding the banner and the modal for a while.

I'm fixing this in bc490b7 by using the `hasReceivedRemotePreferences` selector instead. It doesn't reset back to `false` when preferences are being updated from server. The `AppBanner` component had the same bug, so I'm updating it too. Please check if the changes make sense.

**Review by tyxla — COMMENTED** [review]
> If anyone ever changes the privacy policy version or the notification period, they'll need to toggle the flag back to true.

I'm a bit concerned about that part. How can we be sure that when someone changes it they'll know to come and re-enable the feature flag? Aren't we risking to expose new ToS without notifying the users that way?

**Comment by jsnajdr** [comment]
> How can we be sure that when someone changes it they'll know to come and re-enable the feature flag?

They will need to modify the `/privacy-policy` endpoint implementation, which hardcodes the notification period et al as PHP constants, and when testing that change in Calypso, they will discover that they need to enable the flag. You'd have to completely skip testing in order to miss that.

**Comment by tyxla** [comment]
> You'd have to completely skip testing in order to miss that.

This could theoretically happen if a non-technical person is updating the TOS. I'm still not convinced, but happy to proceed if you'd vouch for this approach as the best one.

**Review by tyxla — APPROVED** [review]
Tests well for me and the code looks good. I wonder if we should eventually completely remove the related code as a next step, especially if we don't see a future use for it anytime soon.

**Comment by jsnajdr** [comment]
> This could theoretically happen if a non-technical person is updating the TOS.

The Privacy Policy (https://automattic.com/privacy/) gets several updates a year (see the changelog at the bottom of the post) but apparently none of them required getting a new consent from users.

> I wonder if we should eventually completely remove the related code

I would be open to just remove it in this very PR if you agree 🙂 It really seems like it won't be used again any time soon.

**Comment by tyxla** [comment]
> I would be open to just remove it in this very PR if you agree 🙂 It really seems like it won't be used again any time soon.

Yep, let's do that. If we ever need to reintroduce it, we can dig it out of the history anyway 😉

**Comment by jsnajdr** [comment]
OK I pushed a complete removal. Now the PR does three things:
- fix the "are preferences loaded?" selector logic in `AppBanner`
- remove the `PrivacyPolicyBanner` and the `privacy-policy` feature flag
- remove the `getCurrentUserRegisterDate` selector that was used only in the `PrivacyPolicyBanner`

How does it look now?

File: client/blocks/app-banner/index.jsx [hunk 1]
```diff
@@ -264,7 +264,7 @@ const mapStateToProps = ( state ) => {
 		dismissedUntil: getPreference( state, APP_BANNER_DISMISS_TIMES_PREFERENCE ),
 		currentSection: getCurrentSection( sectionName, isNotesOpen ),
 		currentRoute,
-		fetchingPreferences: isFetchingPreferences( state ),
+		hasReceivedPreferences: hasReceivedRemotePreferences( state ),
```
Comment (tyxla) [inline client/blocks/app-banner/index.jsx:267]: I wonder if this change deserves its own PR.

**Review by tyxla — APPROVED** [review]
Looking good and working well! 💯 

Would be nice to rename the PR and update the description to reflect what we're actually doing here.

Also, the `AppBanner` fix seems like it belongs in another PR (especially after repurposing this PR), but I'll leave it to you to decide whether to move it out before shipping.

Nice cleanup! 🧹

File: client/blocks/app-banner/index.jsx (hunk 1 above)
Comment (jsnajdr) [reply client/blocks/app-banner/index.jsx:267]: Yes it does deserve its own PR, but now when the PR is just one click away from merging, I'm going to merge it all together.

Comment (tyxla) [reply client/blocks/app-banner/index.jsx:267]: Tempting! Haha
